docs/concepts/trigger-processing.md

```diff
@@ -0,0 +1,55 @@
+
+To deal with this problem, we use the shared leasing mechanism provided by Corvus.Leasing. Under the covers, this is configured to use Azure Blob Storage to aquire and release the leases.
+
+When the engine receives a request to apply a trigger to a workflow instance, it will attempt to acquire a lease for that instance. If the lease can't be acquired (because another process has taken a lease for that instance already), then the engine will retry for up to a minute before failing.
+
+## Trigger processing in a workflow instance
+
+When the workflow engine receives a request to apply a trigger to an instance, it takes the following steps:
+
+1. Acquire a lease for the workflow instance, as described above.
+1. Retrieve the instance from storage
+1. Tell the instance to process the trigger
+    1. Retrieve the workflow for the instance from storage.
+    1. If the instance is faulted or completed, it can't process the trigger, so processing is halted. Note that, as with the remainder of these steps, processing halting due to conditions not being met *does not result in an exception being thrown* because this is a perfectly legitimate outcome.
+    1. Exit conditions for the current state of the instance are evaluated to ensure that the workflow instance can transition out of that state. If any of them evaulate to false, processing is halted.
+    1. Transitions for the current state are evaluated to find a transition that can be executed as a result of the trigger. This is done by iterating the available transitions and choosing the first one whose conditions all evaluate to true. If no transition is found, processing is halted.
+    1. Entry conditions for the transition's target state are evaluated to determine if it is acceptable to move to the new state. If any of the entry conditions evaluate to false, processing is halted.
+    1. Exit actions for the current state are executed.
+    1. Actions for the transition are executed.
+    1. The workflow instance state is updated.
+    1. If the workflow instance is now in a state that has no transitions, it's Status is set to Completed.
+    1. Entry actions for the new state are executed.
+1. If an exception was thrown at any point whilst processing the trigger, put the instance into the Faulted status.
+1. If the instance has been modified, save it back to storage.
+
+There is extensive logging during this process, so if an exception is thrown during processing and the workflow instance does end up in a Faulted state, it should be possible to determine the cause of the failure by reviewing log messages.
+
+However, with that said, it is clear from the above that conditions and actions should only throw exceptions as a last resort. This means a few things.
+
+- Prior to raising a trigger, as much validation as needed should take place so that calling code can be confident that trigger processing will succeed.
```

Comment:
I think we'd need to consider that a bit more before we include it.

If we did what you're suggesting with a trigger, then we have a question to answer. Suppose we're executing a set of 3 actions as part of a transition. The second one fails. If it just throws an exception, the workflow instance is Faulted.

If we want to do something else then we need to avoid Faulting the instance. Therefore we would have to not throw an exception when the second action fails - but as things currently stand there is no way of indicating that it went wrong without the exception, so execution of transition actions would continue, which may not be desirable. At the same time, we'd have queued a trigger to move the instance to our fault handling flow. But the transition that was running continues, and assuming the last action succeeds (as well as entry actions for the new state) then we're relying on the new trigger being processed correctly despite the new state.

This leaves us with a few things we could choose to do:
1. Allow actions to indicate that something has changed that means the currently executing transition needs to be abandoned. I think this could get pretty fiddly, and also doesn't address the case when state entry/exit actions fail. As such, I don't think it's a great idea, because it introduces a new grey area of failure between "everything worked" and "something blew up and you need to fix it" which is "something went wrong but we're relying on something else to take the correct action to sort it out, and we may have left things in an indeterminate state in the meantime."
2. We could suggest that prior to throwing an exception out of an entry/exit action, users should start a new instance of a separate fault handling workflow. The original instance would still be Faulted, but the steps to resolve would be in place.
3. A variation of 2, we could build this functionality into the workflow engine itself. I already wanted to add "global actions" - for example, an action that is called every time a transition executes successfully would be very useful. We could easily add the ability to add actions when a workflow enters the faulted state.

My preference is for no. 3...

Reply:
My reading of @mwadams's comment here was that we would generally advise against designing workflows so that you expect them to get into a faulted state as a matter of course. Error handling is just part of the set of expecting states and transitions.

So faults become things that happen either because something you failed to anticipated has happened, or because there has been some kind of failure that makes it impossible to know what state you're actually in (e.g., your workflow says to invoke some external service as part of an action, and that service is persistently returning 500 errors).

Reply:
Ok, I've added some words that hopefully make this clear...